File: Agent/.NET 6.0/ContextId.cs
using Microsoft.IdentityModel.Tokens;
using System.Reflection.Metadata.Ecma335;
using System.Security.Cryptography;
using System.Text.Json.Serialization;

namespace Technologai
{
    public class ContextId : IComparable<ContextId>
    {
        // hash compute of an id allows to verify which party created this contextId. If that's ever needed.
                
        private readonly string _id;
        private readonly byte[] _unixTimestampBytes = new byte[8];
        private readonly byte[] _hashComputeBytes = new byte[8];

        internal ContextId(ulong unixTimestamp, byte[] idHash)
        {
            _unixTimestampBytes = BitConverter.GetBytes(unixTimestamp);
            _hashComputeBytes = MD5.HashData(idHash.Concat(_unixTimestampBytes).ToArray()).Take(8).ToArray();
            _id = Base64UrlEncoder.Encode(_unixTimestampBytes.Concat(_hashComputeBytes).ToArray());
        }

        [JsonConstructor]
        public ContextId(string contextId)
        {
            _id = contextId;
            var contextBytes = Base64UrlEncoder.DecodeBytes(_id);

            Array.Copy(contextBytes, _unixTimestampBytes, 8);
            Array.Copy(contextBytes, 8, _hashComputeBytes, 0, 8);
        }

        internal static ContextId Create(string creatorIdBase64)        {

            return new ContextId(GetTimestampTicksBytes(), GetBase64Bytes(creatorIdBase64, 8));
        }

        public static ulong GetTimestampTicksBytes()
        {
            return (ulong)(DateTimeOffset.UnixEpoch - DateTimeOffset.UtcNow).Ticks;
        }

        public static byte[] GetBase64Bytes(string creatorIdBase64, int count)
        {
            return Base64UrlEncoder.DecodeBytes(creatorIdBase64).Take(count).ToArray();
        }

        public int CompareTo(ContextId? other)
        {
            if (object.ReferenceEquals(other, null))
            {
                return 1;
            }

            var result = BitConverter.ToInt64(_unixTimestampBytes).CompareTo(
                BitConverter.ToInt64(other._unixTimestampBytes)
            );

            if (result == 0)
            {
                return BitConverter.ToInt64(_hashComputeBytes).CompareTo(
                    BitConverter.ToInt64(other._hashComputeBytes)
                );
            }
            return result;
        }

        public static implicit operator ContextId(string value) => new ContextId(value);

        public static implicit operator string(ContextId value) => value.ToString();       

        public override string ToString()
        {
            return _id;
        }
    }
}

File: Agent/Python/context_id.py
import base64
import hashlib
import time
from dataclasses import dataclass

@dataclass
class ContextId:
    _id: str
    _unix_timestamp_bytes: bytes = None
    _hash_compute_bytes: bytes = None

    def __init__(self, context_id=None, unix_timestamp=None, id_hash=None):
        if context_id:
            self._id = context_id
            context_bytes = base64.urlsafe_b64decode(self._id)
            self._unix_timestamp_bytes = context_bytes[:8]
            self._hash_compute_bytes = context_bytes[8:]
        elif unix_timestamp and id_hash:
            self._unix_timestamp_bytes = unix_timestamp.to_bytes(8, "big")
            self._hash_compute_bytes = hashlib.md5(id_hash + self._unix_timestamp_bytes).digest()[:8]
            self._id = base64.urlsafe_b64encode(self._unix_timestamp_bytes + self._hash_compute_bytes).decode()

    @staticmethod
    def create(creator_id_base64):
        return ContextId(unix_timestamp=ContextId.get_timestamp_ticks_bytes(), id_hash=ContextId.get_base64_bytes(creator_id_base64, 8))

    @staticmethod
    def get_timestamp_ticks_bytes():
        return int(time.time() * 1000)

    @staticmethod
    def get_base64_bytes(creator_id_base64, count):
        return base64.urlsafe_b64decode(creator_id_base64)[:count]

    def compare_to(self, other):
        if other is None:
            return 1

        result = int.from_bytes(self._unix_timestamp_bytes, "big") - int.from_bytes(other._unix_timestamp_bytes, "big")

        if result == 0:
            return int.from_bytes(self._hash_compute_bytes, "big") - int.from_bytes(other._hash_compute_bytes, "big")

        return result

    def __str__(self):
        return self._id

# Example usage
context_id = ContextId.create("your_base64_creator_id")
print(context_id)
